voter/tests/test_fetch.py

```diff
@@ -0,0 +1,412 @@
+        mock_extract.return_value = True
+        mock_listdir.return_value = ['bar.txt', 'ignored.foo']
+        result = voter_fetch_snapshot.process_new_zip(self.url, self.base_path, self.label)
+        self.assertEqual(result, voter_fetch_snapshot.FETCH_STATUS_CODES.CODE_OK)
+        # FileTracker for .txt file gets created ...
+        self.assertTrue(FileTracker.objects.filter(filename='bar.txt').exists())
+        # ... but we ignore files that don't have a .txt extension
+        self.assertFalse(FileTracker.objects.filter(filename='ignored.foo').exists())
+
+    @mock.patch('voter.management.commands.voter_fetch_snapshot.extract_and_remove_file')
+    @mock.patch('voter.management.commands.voter_fetch_snapshot.attempt_fetch_and_write_new_zip')
+    def test_process_new_zip_unable_to_unzip(self, mock_fetch, mock_extract):
+        mock_fetch.return_value = voter_fetch_snapshot.FETCH_STATUS_CODES.CODE_OK, None, None, None
+        mock_extract.return_value = False
+        result = voter_fetch_snapshot.process_new_zip(self.url, self.base_path, self.label)
+        self.assertEqual(result, voter_fetch_snapshot.FETCH_STATUS_CODES.CODE_WRITE_FAILURE)
+
+    @mock.patch('voter.management.commands.voter_fetch_snapshot.attempt_fetch_and_write_new_zip')
+    def test_process_new_zip_already_downloaded(self, mock_fetch):
+        mock_fetch.return_value = voter_fetch_snapshot.FETCH_STATUS_CODES.CODE_NOTHING_TO_DO, None, None, None
+        result = voter_fetch_snapshot.process_new_zip(self.url, self.base_path, self.label)
+        self.assertEqual(result, voter_fetch_snapshot.FETCH_STATUS_CODES.CODE_NOTHING_TO_DO)
+
+    @mock.patch('voter.management.commands.voter_fetch_snapshot.attempt_fetch_and_write_new_zip')
+    def test_process_new_zip_net_failure(self, mock_fetch):
+        mock_fetch.return_value = voter_fetch_snapshot.FETCH_STATUS_CODES.CODE_NET_FAILURE, None, None, None
+        result = voter_fetch_snapshot.process_new_zip(self.url, self.base_path, self.label)
+        self.assertEqual(result, voter_fetch_snapshot.FETCH_STATUS_CODES.CODE_NET_FAILURE)
+
+    @mock.patch('voter.management.commands.voter_fetch_snapshot.attempt_fetch_and_write_new_zip')
+    def test_process_new_zip_write_failure(self, mock_fetch):
+        mock_fetch.return_value = voter_fetch_snapshot.FETCH_STATUS_CODES.CODE_WRITE_FAILURE, None, None, None
+        result = voter_fetch_snapshot.process_new_zip(self.url, self.base_path, self.label)
+        self.assertEqual(result, voter_fetch_snapshot.FETCH_STATUS_CODES.CODE_WRITE_FAILURE)
+
+    # Finally, test the management command itself
+
+    @mock.patch('voter.management.commands.voter_fetch_snapshot.process_new_zip')
+    @mock.patch('voter.management.commands.voter_fetch_snapshot.s3client.list_objects')
+    def test_handle(self, mock_s3_list, mock_process_new_zip):
+        mock_s3_list.return_value = {'Contents': [{'Key': 'http://example.com/foo.zip'}]}
+        call_command('voter_fetch_snapshot')
```

Comment:
I didn't know about `call_command` as a utility for tests!

Reply:
I know! I learned about it when we needed to test some complex commands for Libya. Django is full of gems :)